rudder-license/src/main/scala/com/normation/license/License.scala

```diff
@@ -0,0 +1,724 @@
+  ) extends RudderLicense {
+    def toValid   = ValidSignature  (fileFormat, signature, content, rawSignedContent)
+    def toInvalid = InvalidSignature(fileFormat, signature, content, rawSignedContent)
+  }
+
+  sealed trait CheckedLicense extends RudderLicense
+
+  /**
+   * A licence file for which the signature was checked and is valid.
+   */
+  final case class ValidSignature(
+      fileFormat      : LicenseFileFormat
+    , signature       : LicenseSignature
+    , content         : LicenseInformation
+    , rawSignedContent: String
+  ) extends CheckedLicense
+
+  final case class InvalidSignature(
+      fileFormat      : LicenseFileFormat
+    , signature       : LicenseSignature
+    , content         : LicenseInformation
+    , rawSignedContent: String
+  ) extends CheckedLicense
+}
+
+sealed trait LicenseError {
+  def msg: String
+}
+
+/*
+ * Reason for which a license may be invalid
+ */
+sealed trait InvalidityCause
+final case object InvalidityCause {
+  final case object BadSignature extends InvalidityCause
+  final case class BadSoftwareVersion(current: Version) extends InvalidityCause
+  final case class BadDateTime(current: DateTime) extends InvalidityCause
+}
+
+final object LicenseError {
+  final case class IO       (msg: String)                                                 extends LicenseError
+  final case class Parsing  (msg: String)                                                 extends LicenseError
+  final case class Crypto   (msg: String)                                                 extends LicenseError
+  final case class Signature(msg: String, license: RudderLicense.InvalidSignature)        extends LicenseError
+  final case class Validity (msg: String, cause: InvalidityCause, license: RudderLicense) extends LicenseError
+}
+
+/**
+ * Just a little bit of syntax to be able to tell "...: Maybe[MyType]" in
+ * place of Either[LicenseError, MyType].
+ * And write { ... } withError { ex => LicenseError.Crypto(s"my error is ${ex.getMessage}) }
+ * in place of try { Right(...) } catch { case NonFatal(ex) => Left(...) }
+ */
+final object MaybeLicenseError {
+
+  type Maybe[A] = Either[LicenseError, A]
+
+  //syntax with either
+  final implicit class rightMaybe[A](steps: =>A) {
+    def apply() = Right(steps)
+    def withError(error: Throwable => LicenseError): Maybe[A] = {
+      try { Right(steps) } catch {
+        case NonFatal(ex) => Left(error(ex))
+      }
+    }
+  }
+
+  //syntax with either
+  final implicit class eitherMaybe[A](steps: =>Either[LicenseError, A]) {
+    def apply() = steps
+    def withError(error: Throwable => LicenseError): Maybe[A] = {
+      try { steps } catch {
+        case NonFatal(ex) => Left(error(ex))
+      }
+    }
+  }
+}
+
+/*
+ * An helper object that holds ISO date reader/writer
+ */
+final object ISODate {
+  import MaybeLicenseError._
+
+  val pattern = "yyyy-MM-dd'T'HH:mm:ssZZ" // because no way to get back from formatter to pattern...
```

Comment:
🤦‍♂️

Reply:
I think I should use the pattern as an entry point and don't use the provided ISO formatter at all, so that we can't have a desynchronisation between the two of them (there is no perf cost, JodaTime is caching the thing internally).